Boundary Docs: Configure workers refactor (#3077)
hashicorp/web-unified-docs: content/boundary/v1.0.x/content/docs/workers/troubleshoot.mdx

---
layout: docs
page_title: Troubleshoot workers
description: >-
  Diagnose Boundary worker problems. Resolve upstream connection failures, registration errors, worker filter mismatches, and session recording storage issues.
# START AUTO GENERATED METADATA, DO NOT EDIT
created_at: 2026-08-10T22:18:19.000Z
last_modified: 2026-08-10T22:18:19.000Z
# END AUTO GENERATED METADATA
---

# Troubleshoot workers

Boundary workers fail in a small number of recognizable ways.
A worker cannot reach its upstream, it never finishes registering, a worker filter excludes it from a session, or it runs out of local storage.
Each of the following symptoms includes the exact error Boundary returns, its cause, and the resolution.

> ## Collect diagnostic information

Reviewer: Since these diagnostic sources are all optional and you wouldn't necessarily do them consecutively, I think they would make more sense under H3 headings instead of numbered steps.

Author: agreed, commit 0e8ccc45f09cfd0ff8377bdf4cdf988726060b17 changes this to use H3 headers

   ```

   A stale `Last Status Time` value means the worker stopped reporting to its upstream.

1. The contents of the worker's `auth_storage_path` directory:

   ```shell-session
   $ ls /var/lib/boundary
   ```

   A worker that registered has a `nodecreds` directory.
   A worker that has not registered also has an `auth_request_token` file.

1. The worker's health endpoint, if you configured the worker with a listener that has the `ops` purpose:

   ```shell-session
   $ curl -s "http://localhost:9203/health?worker_info=1"
   ```

   The response reports whether the worker can reach its upstream:

   <CodeBlockConfig hideClipboard>

   ```plaintext
   {"worker_process_info":{"state":"active","active_session_count":0,"upstream_connection_state":"READY"}}
   ```

   </CodeBlockConfig>

   An `upstream_connection_state` value of `READY` means the worker is connected. Any other value, such as `TRANSIENT_FAILURE`, means it is not.

   <Warning>

     The worker health endpoint always returns `200 OK`, even when the worker cannot reach its upstream.
     Check the `upstream_connection_state` field in the response body rather than the HTTP status code.

>    </Warning>

Reviewer: ```suggestion
   <Note>

     The worker health endpoint always returns `200 OK`, even when the worker cannot reach its upstream.
     Check the `upstream_connection_state` field in the response body rather than the HTTP status code.

   </Note>
```

Reviewer: I think this should be a note since it's additional info the user may need to take and not a breaking change or security threat.

**Resolution:** Find and stop the process that holds the port, then start the worker again:

```shell-session
$ sudo lsof -i :9202
```

Confirm that the earlier worker exited.
A worker in graceful shutdown continues to hold its port until its sessions drain or you send a second interrupt.

## The worker cannot record sessions

**Symptom:** Boundary does not select the worker to record sessions, or session recording fails.

**Cause:** The worker's local storage state is not `available`, or the worker cannot reach the external storage.

**Resolution:** Read the worker and check the `Local Storage State` field:

```shell-session
$ boundary workers read -id w_pkx59YrH6g
```

The following values indicate a problem:

- `not configured` - The worker has no `recording_storage_path` value. Add the parameter and restart the worker.
- `low storage` - Available disk space fell below the `recording_storage_minimum_available_capacity` value.
- `critically low storage` - Available disk space fell below half the `recording_storage_minimum_available_capacity` value.
- `out of storage` - The worker has less than 1 MB of available disk space.

> To recover, free disk space on the worker.

Reviewer: ```suggestion
To resolve the issue, free disk space on the worker.
```

Reviewer: Recover might be confusing here since we have the session recording recovery feature now.

The worker returns to the `available` state on its own, without a restart.

<Note>

  A worker in an unhealthy storage state does not return to `available` as soon as it crosses back over the threshold.
  Available disk space must reach 125 percent of the `recording_storage_minimum_available_capacity` value.
  Boundary requires the extra headroom so that a worker near the threshold does not flip back and forth between states.

> </Note>

Reviewer: ```suggestion

A worker in an unhealthy storage state does not return to `available` as soon as it crosses back over the threshold.
  Available disk space must reach 125 percent of the `recording_storage_minimum_available_capacity` value.
  Boundary requires the extra disk space so that a worker near the threshold does not switch back and forth between states.

```

Reviewer: Since this is just informative, it shouldn't be a note.
